crates/types/src/blockchain/header.rs

```diff
@@ -37,6 +47,58 @@ pub struct BlockHeader {
     metadata: Option<BlockHeaderMetadata>,
 }
 
+impl From<BlockHeaderV1> for BlockHeader {
+    fn from(v1: BlockHeaderV1) -> Self {
+        BlockHeader::V1(v1)
+    }
+}
+
+impl BlockHeader {
+    /// Getter for consensus portion of header
+    pub fn consensus(&self) -> &ConsensusHeader<GeneratedConsensusFields> {
+        match self {
+            BlockHeader::V1(v1) => &v1.consensus,
+        }
+    }
+
+    /// Getter for application portion of header
+    pub fn application(&self) -> &ApplicationHeader<GeneratedApplicationFields> {
+        match self {
+            BlockHeader::V1(v1) => &v1.application,
+        }
+    }
+
+    /// Getter for metadata portion of header
+    pub fn metadata(&self) -> &Option<BlockHeaderMetadata> {
+        match self {
+            BlockHeader::V1(v1) => &v1.metadata,
+        }
+    }
+
+    /// Mutable getter for consensus portion of header
+    pub fn consensus_mut(&mut self) -> &mut ConsensusHeader<GeneratedConsensusFields> {
+        match self {
+            BlockHeader::V1(v1) => &mut v1.consensus,
+        }
+    }
+
+    /// Mutable getter for application portion of header
+    pub fn application_mut(
+        &mut self,
+    ) -> &mut ApplicationHeader<GeneratedApplicationFields> {
+        match self {
+            BlockHeader::V1(v1) => &mut v1.application,
+        }
+    }
+
+    /// Mutable getter for metadata portion of header
+    pub fn metadata_mut(&mut self) -> &mut Option<BlockHeaderMetadata> {
```

Comment:
And we definitely don't want to allow modification of the metadata. You use that only in one place and you can work directly with enum there.